File: memories/2024/exercise/scala/day13/src/test/scala/santamarket/model/SantamarketTest.scala
package santamarket.model

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class SantamarketTest extends AnyFunSuite with Matchers {

  test("no discount") {
    val catalog = new FakeCatalog()
    val teddyBear = Product("teddyBear", ProductUnit.Each)
    catalog.addProduct(teddyBear, 1.0)

    val sleigh = new ShoppingSleigh()
    sleigh.addItemQuantity(teddyBear, 3.0)

    val elf = new ChristmasElf(catalog)
    val receipt = elf.checksOutArticlesFrom(sleigh)

    receipt.getTotalPrice should equal(3.0 +- 0.01)
  }

  test("ten percent discount") {
    val catalog = new FakeCatalog()
    val turkey = Product("turkey", ProductUnit.Kilo)
    catalog.addProduct(turkey, 2.0)

    val sleigh = new ShoppingSleigh()
    sleigh.addItemQuantity(turkey, 2.0)

    val elf = new ChristmasElf(catalog)
    elf.addSpecialOffer(SpecialOfferType.TenPercentDiscount, turkey, 10.0)
    val receipt = elf.checksOutArticlesFrom(sleigh)

    receipt.getTotalPrice should equal(3.6 +- 0.01) // 4.0 - 0.4
  }

  test("three for two discount") {
    val catalog = new FakeCatalog()
    val teddyBear = Product("teddyBear", ProductUnit.Each)
    catalog.addProduct(teddyBear, 1.0)

    val elf = new ChristmasElf(catalog)
    elf.addSpecialOffer(SpecialOfferType.ThreeForTwo, teddyBear, 0.0)

    val sleigh = new ShoppingSleigh()
    sleigh.addItemQuantity(teddyBear, 3.0)

    val receipt = elf.checksOutArticlesFrom(sleigh)

    receipt.getTotalPrice should equal(2.0 +- 0.01) // Buy 3, pay for 2
  }
}

File: memories/2024/exercise/C#/day13/SantaMarket.Tests/SantamarketTest.cs
using Xunit;

namespace SantaMarket.Model.Tests
{
    public class SantamarketTest
    {
        [Fact]
        public void NoDiscount()
        {
            var catalog = new FakeCatalog();
            var teddyBear = new Product("teddyBear", ProductUnit.Each);
            catalog.AddProduct(teddyBear, 1.0);

            var sleigh = new ShoppingSleigh();
            sleigh.AddItemQuantity(teddyBear, 3);

            var elf = new ChristmasElf(catalog);
            var receipt = elf.ChecksOutArticlesFrom(sleigh);

            Assert.Equal(3.0, receipt.GetTotalPrice(), 2);
        }

        [Fact]
        public void TenPercentDiscount()
        {
            var catalog = new FakeCatalog();
            var turkey = new Product("turkey", ProductUnit.Kilo);
            catalog.AddProduct(turkey, 2.0);

            var sleigh = new ShoppingSleigh();
            sleigh.AddItemQuantity(turkey, 2.0);

            var elf = new ChristmasElf(catalog);
            elf.AddSpecialOffer(SpecialOfferType.TenPercentDiscount, turkey, 10.0);
            var receipt = elf.ChecksOutArticlesFrom(sleigh);

            Assert.Equal(3.6, receipt.GetTotalPrice(), 2);  // 4.0 - 0.4
        }

        [Fact]
        public void ThreeForTwoDiscount()
        {
            var catalog = new FakeCatalog();
            var teddyBear = new Product("teddyBear", ProductUnit.Each);
            catalog.AddProduct(teddyBear, 1.0);

            var elf = new ChristmasElf(catalog);
            elf.AddSpecialOffer(SpecialOfferType.ThreeForTwo, teddyBear, 0.0);

            var sleigh = new ShoppingSleigh();
            sleigh.AddItemQuantity(teddyBear, 3);

            var receipt = elf.ChecksOutArticlesFrom(sleigh);

            Assert.Equal(2.0, receipt.GetTotalPrice(), 2);  // Buy 3, pay for 2
        }
    }
}
